lib/archive/util.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2016-2017 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package archive
+
+import (
+	"fmt"
+	"path/filepath"
+	"strings"
+)
+
+const (
+	// CopyTo is used to indicate that the desired filter spec is for a CopyTo direction
+	CopyTo = true
+
+	// CopyFrom is used to indicate that the desired filter spec is for the CopyFrom direction
+	CopyFrom = false
+)
+
+// GenerateFilterSpec will populate the appropriate relative Rebase and Strip paths based on the supplied scenarios. Inclusion/Exclusion should be constructed separately. Please also note that any mount that exists before the copy target that is not primary and comes before the primary target will have a bogus filterspec, since it would not be written or read to.
+func GenerateFilterSpec(copyPath string, mountPoint string, primaryTarget bool, direction bool) FilterSpec {
```

Comment:
For my understanding, this filter spec is requited and becomes more and more complicated, is because we're not mounting everything one time, and do one diff for the whole file system like what docker did. In that way, we could just have one rebase path, and everything works.
But we have to mount container r/w layer, and volumes one after another. If we're copying to multiple devices, with same parent mount point, the strip and rebase becomes different for different device.
And for the same reason, the primaryTarget is invented, just is used to generate strip and rebase string. This variable is not used by the filterSpec consumer, and could be derived from the mount points, copyTo/From target and file path.

So here my suggestion is that, change the signature to accept copySource, copyTarget, direction, and all the mount points, instead of only one, and return one filterspec per mount point.

In this method, the inclusion/exclusion should be filled based on the input. If there could have some like the strip slash, etc.

If user have their own inclusion/exclusion specification, this method could have that information as input as well, and fill the Inclusion/Exclusion in FilterSpec.

We should not have the caller aware of too many internal details, e.g. PrimaryTarget, to generate one filter spec.